src/expr/src/scalar/func.rs

```diff
@@ -3360,6 +3364,23 @@ fn array_lower<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
     }
 }
 
+fn array_upper<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
```

Comment:
This seems to have a similar issue as #4517; the second argument expresses the dimension you want this metric from, and it looks like this is instead using `i` has the "horizontal" iterator across the array, e.g. this code could not match PG's behavior with an `i` value of 3 on e.g. `ARRAY[[[1]]]`.

Reply:
I believe something like this should work! (As mentioned in #4517, these sorts of functions are much easier to write for arrays.)

```rust
match a.unwrap_array().dims().into_iter().nth(i) {
    Some(dim) => dim.length,
    None => Datum::Null,
}
```